bin/src/upgrade.rs

```diff
@@ -54,17 +55,20 @@ pub struct UpgradeData {
   //pub token_count: usize,
 }
 
-pub fn start_new_main_process(executable_path: String, upgrade_data: UpgradeData) -> (pid_t, Channel<(),bool>) {
+pub fn start_new_main_process(
+  executable_path: String,
+  upgrade_data: UpgradeData
+) -> Result<(pid_t, Channel<(),bool>), anyhow::Error> {
   trace!("parent({})", unsafe { libc::getpid() });
 
-  let mut upgrade_file = tempfile().expect("could not create temporary file for upgrade");
+  let mut upgrade_file = tempfile().context("could not create temporary file for upgrade")?;
 
   util::disable_close_on_exec(upgrade_file.as_raw_fd());
 
-  serde_json::to_writer(&mut upgrade_file, &upgrade_data).expect("could not write upgrade data to temporary file");
-  upgrade_file.seek(SeekFrom::Start(0)).expect("could not seek to beginning of file");
+  serde_json::to_writer(&mut upgrade_file, &upgrade_data).context("could not write upgrade data to temporary file")?;
```

Comment:
if you could switch from the context method to the with_context method, it will help to not allocate each time the error, even though it is a success